feat: added accessor protocol support to `stats/base/meanpn`

dhruvilmehta opened the pull request:
Resolves #5650 

## Description

> What is the purpose of this pull request?

This pull request:

-   Added accessor protocol to meanpn.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5650 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [X] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

dhruvilmehta committed:
fix: fixes given comments on the given commit

Closes: #5718

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

dhruvilmehta committed:
feat: added accessor protocol for meanpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

dhruvilmehta committed:
fix: fixed coyright years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

dhruvilmehta committed:
fix: fixed coyright years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte committed:
Discard changes to lib/node_modules/@stdlib/stats/base/dnanvariancech/README.md

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
```diff
@@ -67,9 +67,9 @@ The function has the following parameters:
+-   **strideX**: stride length for `x`.
 
-The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the [arithmetic mean][arithmetic-mean] of every other element in `x`,
+The `N` and `strideX` parameters determine which elements in `x` are accessed at runtime. For example, to compute the [arithmetic mean][arithmetic-mean] of every other element in `x`,
```
```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the [arithmetic mean][arithmetic-mean] of every other element in the input array,
```

dhruvilmehta committed:
Update lib/node_modules/@stdlib/stats/base/meanpn/README.md

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Dhruvil Mehta <68022411+dhruvilmehta@users.noreply.github.com>

dhruvilmehta commented:
@kgryte That was the only change needed, right? Let me know of any other changes.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
```diff
@@ -51,7 +51,7 @@ The [arithmetic mean][arithmetic-mean] is defined as
-#### meanpn( N, x, stride )
+#### meanpn( N, x, strideX )
 
 Computes the [arithmetic mean][arithmetic-mean] of a strided array `x` using a two-pass error correction algorithm.
```
```suggestion
Computes the [arithmetic mean][arithmetic-mean] of a strided array using a two-pass error correction algorithm.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
````diff
@@ -67,9 +67,9 @@ The function has the following parameters:
+The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the [arithmetic mean][arithmetic-mean] of every other element in the input array,
 
 ```javascript
 var floor = require( '@stdlib/math/base/special/floor' );
````
@dhruvilmehta Would you also mind updating the examples throughout this package to no longer use `floor`?

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
```diff
@@ -112,7 +112,7 @@ var v = meanpn.ndarray( N, x, 1, 0 );
+-   **offsetX**: starting index for `x`.
 
 While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other value in `x` starting from the second value
 
```
See other similar packages, but we also tend to add a note concerning accessor protocol support below.

kgryte committed:
Discard changes to lib/node_modules/@stdlib/stats/base/meanpn/benchmark/benchmark.ndarray.js

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/docs/repl.txt:
```diff
@@ -49,7 +49,8 @@
     > {{alias}}( N, x1, stride )
     ~-0.3333
 
-{{alias}}.ndarray( N, x, stride, offset )
```
Need to factor out the use of `floor` above.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/docs/types/index.d.ts:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2020 The Stdlib Authors.
+* Copyright (c) 2025 The Stdlib Authors.
```
In general, unless you are adding new files, you don't need to update the copyright years of each file you touch.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/examples/index.js:
```diff
@@ -18,18 +18,12 @@
-for ( i = 0; i < x.length; i++ ) {
-	x[ i ] = round( (randu()*100.0) - 50.0 );
-}
+var x = discreteUniform( 10, 0, 100, {
```
```suggestion
var x = discreteUniform( 10, -50, 50, {
```

Preserve the original value range. Also, you need to update the example in the README, as well, to match the changes you made here.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,77 @@
+/**
+* Computes the arithmetic mean of a strided array using a two-pass error correction algorithm.
+*
+* @param {PositiveInteger} N - number of indexed elements
```
```suggestion
* @private
* @param {PositiveInteger} N - number of indexed elements
```

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,77 @@
+* Computes the arithmetic mean of a strided array using a two-pass error correction algorithm.
+*
+* @param {PositiveInteger} N - number of indexed elements
+* @param {NumericArray} x - input array
```
This type and description isn't correct. You are expecting an `input array object`.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,77 @@
+* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
+* var arraylike2object = require( '@stdlib/array/base/arraylike2object' );
+*
+* var x = toAccessorArray([ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
```
Spacing is off here and below.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/main.js:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2020 The Stdlib Authors.
+* Copyright (c) 2025 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2020 The Stdlib Authors.
```

You didn't touch this file, so why are you updating the copyright year?

kgryte committed:
Discard changes to lib/node_modules/@stdlib/stats/base/meanpn/test/test.js

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/test/test.meanpn.js:
```diff
@@ -40,6 +41,23 @@ tape( 'the function has an arity of 3', function test( t ) {
+
+	x = [ 1.0, -2.0, -4.0, 5.0, 0.0, 3.0 ];
+
+	v = meanpn( x.length, toAccessorArray(x), 1 );
```
```suggestion
	v = meanpn( x.length, toAccessorArray( x ), 1 );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/test/test.meanpn.js:
```diff
@@ -40,6 +41,23 @@ tape( 'the function has an arity of 3', function test( t ) {
+	t.end();
+});
+
 tape( 'the function calculates the arithmetic mean of a strided array', function test( t ) {
```
You need to add explicit test blocks below which test accessor behavior just as we do for indexed arrays. You did one test block, but you are missing all the others for L61 and below.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/test/test.ndarray.js:
```diff
@@ -56,6 +57,23 @@ tape( 'the function calculates the arithmetic mean of a strided array', function
 	t.end();
 });
 
+tape( 'the function calculates the arithmetic mean of a strided array (accessors)', function test( t ) {
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/test/test.ndarray.js:
```diff
@@ -56,6 +57,23 @@ tape( 'the function calculates the arithmetic mean of a strided array', function
+
+	x = [ 1.0, -2.0, -4.0, 5.0, 0.0, 3.0 ];
+
+	v = meanpn( x.length, toAccessorArray(x), 1, 0 );
```
```suggestion
	v = meanpn( x.length, toAccessorArray( x ), 1, 0 );
```

Here and below.

dhruvilmehta committed:
fix: made suggested changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

dhruvilmehta commented:
Done with the suggested changes and resolved the conversations accordingly.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
```diff
@@ -87,42 +82,35 @@ Note that indexing is relative to the first index. To introduce an offset, use [
+-   **offsetX**: starting index for `x`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other value in the strided array starting from the second element
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other element in the strided array starting from the second element
```

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
```diff
@@ -135,6 +123,7 @@ var v = meanpn.ndarray( N, x, 2, 1 );
+-   Both functions support array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).
 -   Depending on the environment, the typed versions ([`dmeanpn`][@stdlib/stats/base/dmeanpn], [`smeanpn`][@stdlib/stats/base/smeanpn], etc.) are likely to be significantly more performant.
 
 </section>
```
The example below still needs updating.

kgryte replied:
See `examples/index.js`.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/docs/repl.txt:
```diff
@@ -32,24 +32,21 @@
     > var x = [ 1.0, -2.0, 2.0 ];
     > {{alias}}( x.length, x, 1 )
-    ~0.3333
+    0.3333333333333333
```
This change is not needed.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/docs/repl.txt:
```diff
@@ -32,24 +32,21 @@
-    > {{alias}}( N, x, stride )
-    ~0.3333
+    > {{alias}}( 3, x, 2 )
+    0.3333333333333333
```
This change is not needed.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/docs/repl.txt:
```diff
@@ -32,24 +32,21 @@
-    > stride = 2;
-    > {{alias}}( N, x1, stride )
-    ~-0.3333
+    > {{alias}}( 6, x1, 2 )
```
The `N` argument is not correct. L46 is not correct.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,75 @@
+
+/**
+* Computes the arithmetic mean of a strided array using a two-pass error correction algorithm.
+* @private
```
```suggestion
*
* @private
```

Always include a blank line between the description and tags.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,75 @@
+	var mu;
+	var c;
+
+	xbuf = x.data;
+
+	get = x.accessors[ 0 ];
+
+	if ( N === 1 || strideX === 0 ) {
+		return get( xbuf, offsetX );
+	}
+
+	ix = offsetX;
```
```suggestion
	xbuf = x.data;
	get = x.accessors[ 0 ];

	if ( N === 1 || strideX === 0 ) {
		return get( xbuf, offsetX );
	}
	ix = offsetX;
```

kgryte replied:
Why is the `ix` variable even needed? Can you not use `offsetX` in the calls below?

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/index.js:
```diff
@@ -33,13 +33,11 @@
 * // returns ~0.3333
```
You can remove the `N` declaration above and inline the argument.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/meanpn.js:
```diff
@@ -50,23 +50,8 @@ var gapxsumpw = require( '@stdlib/blas/ext/base/gapxsumpw' );
 * var v = meanpn( N, x, 1 );
```
You can remove the `N` declaration above and inline the argument value.

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/test/test.ndarray.js:
```diff
@@ -141,8 +196,19 @@ tape( 'if provided a `stride` parameter equal to `0`, the function returns the f
+	t.end();
+});
+
 tape( 'the function supports an `offset` parameter', function test( t ) {
```
Is there a reason you did not include a similar test for accessor arrays?

dhruvilmehta replied:
Sorry I missed that. Included in the next commit.

dhruvilmehta committed:
fix: made suggested changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
```diff
@@ -87,42 +82,35 @@ Note that indexing is relative to the first index. To introduce an offset, use [
+-   **offsetX**: starting index for `x`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other element in the strided array starting from the second element
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other element in the strided array starting from the second element
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/README.md:
```diff
@@ -151,15 +140,12 @@ var v = meanpn.ndarray( N, x, 2, 1 );
 var randu = require( '@stdlib/random/base/randu' );
 var round = require( '@stdlib/math/base/special/round' );
 var Float64Array = require( '@stdlib/array/float64' );
```
```suggestion
```

kgryte committed:
docs: remove unused imports in example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,72 @@
+*
+* @private
+* @param {PositiveInteger} N - number of indexed elements
+* @param {NumericArray} x - input array object
```
```suggestion
* @param {Object} x - input array object
```

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,72 @@
+	if ( N === 1 || strideX === 0 ) {
+		return get( xbuf, offsetX );
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/meanpn/lib/accessors.js:
```diff
@@ -0,0 +1,72 @@
+
+	// Compute an error term:
+	c = gapxsumpw( N, -mu, xbuf, strideX, offsetX ) / N;
+	return mu + c;
```
```suggestion

	return mu + c;
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up.
